src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Linux.cs

```diff
@@ -0,0 +1,1200 @@
+                return errorCode;
+            }
+
+            var operation = new ConnectOperation {
+                Callback = callback,
+                SocketAddress = socketAddress,
+                SocketAddressLen = socketAddressLen
+            };
+
+            bool isStopped;
+            while (!TryBeginOperation(ref _acceptOrConnectQueue, operation, out isStopped))
+            {
+                if (isStopped)
+                {
+                    // TODO: is this error code reasonable for a closed socket? Check with Winsock.
+                    operation.ErrorCode = SocketError.Shutdown;
+                    QueueCompletion(operation);
+                    return SocketError.Shutdown;
+                }
+
+                if (TryCompleteConnect(_fileDescriptor, operation))
+                {
+                    QueueCompletion(operation);
+                    break;
+                }
+            }
+            return SocketError.IOPending;
+        }
+
+        private static bool TryCompleteConnect(int fileDescriptor, ConnectOperation operation)
+        {
+            return TryCompleteConnect(fileDescriptor, operation.SocketAddress, operation.SocketAddressLen, out operation.ErrorCode);
+        }
+
+        private static unsafe bool TryCompleteConnect(int fileDescriptor, byte[] socketAddress, int socketAddressLen, out SocketError errorCode)
+        {
+            Debug.Assert(socketAddress != null);
+            Debug.Assert(socketAddressLen > 0);
+
+            int err;
+            fixed (byte* rawSocketAddress = socketAddress)
+            {
+                var sockAddr = (Interop.libc.sockaddr*)rawSocketAddress;
+                err = Interop.libc.connect(fileDescriptor, sockAddr, (uint)socketAddressLen);
+            }
+
+            if (err == 0)
+            {
+                errorCode = SocketError.Success;
+                return true;
+            }
+
+            Interop.Error errno = Interop.Sys.GetLastError();
+            if (errno != Interop.Error.EINPROGRESS)
+            {
+                errorCode = SocketPal.GetSocketErrorForErrorCode(errno);
+                return true;
+            }
+
+            errorCode = SocketError.Success;
+            return false;
+        }
+
+        private static unsafe int Receive(int fd, int flags, int available, byte[] buffer, int offset, int count, byte[] socketAddress, ref int socketAddressLen, out int receivedFlags, out Interop.Error errno)
+        {
+            Debug.Assert(socketAddress != null || socketAddressLen == 0);
+
+            var pinnedSocketAddress = default(GCHandle);
+            Interop.libc.sockaddr* sockAddr = null;
+            uint sockAddrLen = 0;
+
+            int received;
+            try
+            {
+                if (socketAddress != null)
+                {
+                    pinnedSocketAddress = GCHandle.Alloc(socketAddress, GCHandleType.Pinned);
```

Comment:
Does `fixed` work when the object being pinned is `null`? The C# spec leaves the behavior up to the implementation (see section 18.6 of the C# 5.0 spec: "The behavior of the fixed statement is implementation-defined if the array expression is null or if the array has zero elements.").